gossipd/routing.c

```diff
@@ -332,10 +333,30 @@ static u64 risk_fee(u64 amount, u32 delay, double riskfactor)
 
 /* We track totals, rather than costs.  That's because the fee depends
  * on the current amount passing through. */
-static void bfg_one_edge(struct node *node, size_t edgenum, double riskfactor)
+static void bfg_one_edge(struct node *node, size_t edgenum, double riskfactor,
+			 double fuzz, const struct isaac64_ctx *baserng)
 {
 	struct node_connection *c = node->in[edgenum];
 	size_t h;
+	struct isaac64_ctx myrng;
+	double fee_scale = 1.0;
+	u8 *scid;
+
+	if (fuzz != 0.0) {
+		/* Copy RNG state */
+		myrng = *baserng;
+
+		/* Provide the channel short ID, as additional
+		 * entropy. */
+		scid = tal_arr(NULL, u8, 0);
+		towire_short_channel_id(&scid, &c->short_channel_id);
+		isaac64_reseed(&myrng,
+			       (const unsigned char *) scid, tal_len(scid));
+		tal_free(scid);
```

Comment:
That's pretty heavy.  It's nice because it makes the per-edge decision deterministic, though.

`run-bench-find_route 10000 10` with 0 fuzz:

    10 (10 succeeded) routes in 10000 nodes in 3990 msec (399066864 nanoseconds per route)

With 0.75 fuzz and a fixed rng:

    10 (10 succeeded) routes in 10000 nodes in 73423 msec (7342349728 nanoseconds per route)

Ouch!

I switched it to hand in a 64-bit base_seed, and use short_channel_id_to_uint() instead of wire allocation, and initialize the rng with [base_seed, short_channel_id_to_uint()], but it's still v. slow:

    10 (10 succeeded) routes in 10000 nodes in 67433 msec (6743351288 nanoseconds per route)

This is partially because it's expensive to set up an isaac64 context just to pull one number out; we're better using siphash for this:

    10 (10 succeeded) routes in 10000 nodes in 5668 msec (566855365 nanoseconds per route)

I've put a rough patch (enough to compile and run the bench): https://gist.github.com/rustyrussell/c3334112c096d97e0adb0d3a8fc3fc32

Reply:
The patch is actually pretty good and I would ACK that.